### ISSUE TITLE
ML matlab implementation question

### ISSUE THREAD

**Issue by aldipiroli** [gledsonmelotti/ML-MAP-Layers-for-Probabilistic#1]
Hello and thanks for sharing your work! 

I am trying to understand the matlab code that you provided for the ML score. 
My question comes from these [lines](https://github.com/gledsonmelotti/ML-MAP-Layers-for-Probabilistic/blob/583ea979d66f071c6acf7ddae74a69f0a315e0ba/ML_MAP_Layer_General/ML_ObjSc.m#L79-L87):

```
ObjnessCar_TP = Car_TP_ObjSc;
ObjnessCyc_TP = Cyc_TP_ObjSc;
ObjnessPed_TP = Ped_TP_ObjSc;


ML_TP = P+lambda;
ML_TP = ML_TP./sum(ML_TP,2);
ML_TP(1:size(ObjnessCar_TP,1),1) = ML_TP(1:size(ObjnessCar_TP,1),1).*ObjnessCar_TP;
ML_TP(size(ObjnessCar_TP,1)+1:size(ObjnessCar_TP,1)+size(ObjnessCyc_TP,1),2) = ML_TP(size(ObjnessCar_TP,1)+1:size(ObjnessCar_TP,1)+size(ObjnessCyc_TP,1),2).*ObjnessCyc_TP;
ML_TP(size(ObjnessCar_TP,1)+size(ObjnessCyc_TP,1)+1:size(ObjnessCar_TP,1)+size(ObjnessCyc_TP,1)+size(ObjnessPed_TP,1),3) = ML_TP(size(ObjnessCar_TP,1)+size(ObjnessCyc_TP,1)+1:size(ObjnessCar_TP,1)+size(ObjnessCyc_TP,1)+size(ObjnessPed_TP,1),3).*ObjnessPed_TP;
``` 

I understand how the ML_TP score is computed, `ML_TP = ML_TP./sum(ML_TP,2);`, but I do not understand what is the goal of the successive lines.
From the paper I understand that the final ML score  is derived as `ML_TP = ML_TP./sum(ML_TP,2);`, but in the code the final value is also multiplied by `ObjnessCar_TP`, which I don't understand its meaning.

Also, from a quick look, it seems that `ClS_Test_TP = [Car_TP_raw_sc;Cyc_TP_raw_sc;Ped_TP_raw_sc];` is loading the scores from the training data, and not the test data. `ClS_Test_TP` should represent the raw logits scores from the test predictions right?
 

Thanks again.

**Comment by gledsonmelotti** [comment]
> Hello and thanks for sharing your work!
> 
> I am trying to understand the matlab code that you provided for the ML score. My question comes from these [lines](https://github.com/gledsonmelotti/ML-MAP-Layers-for-Probabilistic/blob/583ea979d66f071c6acf7ddae74a69f0a315e0ba/ML_MAP_Layer_General/ML_ObjSc.m#L79-L87):
> 
> ```
> ObjnessCar_TP = Car_TP_ObjSc;
> ObjnessCyc_TP = Cyc_TP_ObjSc;
> ObjnessPed_TP = Ped_TP_ObjSc;
> 
> 
> ML_TP = P+lambda;
> ML_TP = ML_TP./sum(ML_TP,2);
> ML_TP(1:size(ObjnessCar_TP,1),1) = ML_TP(1:size(ObjnessCar_TP,1),1).*ObjnessCar_TP;
> ML_TP(size(ObjnessCar_TP,1)+1:size(ObjnessCar_TP,1)+size(ObjnessCyc_TP,1),2) = ML_TP(size(ObjnessCar_TP,1)+1:size(ObjnessCar_TP,1)+size(ObjnessCyc_TP,1),2).*ObjnessCyc_TP;
> ML_TP(size(ObjnessCar_TP,1)+size(ObjnessCyc_TP,1)+1:size(ObjnessCar_TP,1)+size(ObjnessCyc_TP,1)+size(ObjnessPed_TP,1),3) = ML_TP(size(ObjnessCar_TP,1)+size(ObjnessCyc_TP,1)+1:size(ObjnessCar_TP,1)+size(ObjnessCyc_TP,1)+size(ObjnessPed_TP,1),3).*ObjnessPed_TP;
> ```
> 
> I understand how the ML_TP score is computed, `ML_TP = ML_TP./sum(ML_TP,2);`, but I do not understand what is the goal of the successive lines. From the paper I understand that the final ML score is derived as `ML_TP = ML_TP./sum(ML_TP,2);`, but in the code the final value is also multiplied by `ObjnessCar_TP`, which I don't understand its meaning.
> 
> Also, from a quick look, it seems that `ClS_Test_TP = [Car_TP_raw_sc;Cyc_TP_raw_sc;Ped_TP_raw_sc];` is loading the scores from the training data, and not the test data. `ClS_Test_TP` should represent the raw logits scores from the test predictions right?
> 
> Thanks again.

Hi, how are you? I followed the Yolo V4 methodology, which multiplies the final classification scores by the objectness scores. But this is not mandatory. Some detectors do not have objectness scores. The referenced lines are for allocating the objectness scores correctly from the class scores.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
Take a look at an example of classification.

diretorio_probability = '.../Logits/';
load([diretorio_probability,'Probability_logits_train.mat']);
load([diretorio_probability,'Train_labels.mat']);

load([diretorio_probability,'Probability_logits_test.mat']);
load([diretorio_probability,'Test_labels.mat']);
load([diretorio_probability,'Test_predict.mat']);
load([diretorio_probability,'Probability_test.mat']);

c = unique(Train_labels);
nc = length(c);
ctr1 = Train_labels == c(1);
ctr2 = Train_labels == c(2);
ctr3 = Train_labels == c(3);

par_ml=[];

nbins_car=8;
nbins_cyc=18;
nbins_ped=28;

figure(1);
set(gcf,'color','w');grid
cla; grid; title('Logits Dist. on Training'); hold on
hc1 = histogram(Probability_logits_train(ctr1,1),nbins_car,'Normalization','probability');
hc2 = histogram(Probability_logits_train(ctr2,2),nbins_cyc,'Normalization','probability');
hc3 = histogram(Probability_logits_train(ctr3,3),nbins_ped,'Normalization','probability');

Valores1 = hc1.Values;
BinEdgesLow1 = hc1.BinEdges(1:nbins_car);
BinEdgesHigh1 = hc1.BinEdges(2:nbins_car+1);

Valores2 = hc2.Values;
BinEdgesLow2 = hc2.BinEdges(1:nbins_cyc);
BinEdgesHigh2 = hc2.BinEdges(2:nbins_cyc+1);

Valores3 = hc3.Values;
BinEdgesLow3 = hc3.BinEdges(1:nbins_ped);
BinEdgesHigh3 = hc3.BinEdges(2:nbins_ped+1);

Y = Probability_logits_test;
n = size(Y,1);

P1 = zeros(n,1);
P2 = zeros(n,1);
P3 = zeros(n,1);
for k=1:n
    for i=1:size(Valores1,2)
        if (BinEdgesLow1(i) <= Y(k,1)) && (Y(k,1) < BinEdgesHigh1(i))
            P1(k) = Valores1(i);
        end
    end
    for i=1:size(Valores2,2)
        if (BinEdgesLow2(i) <= Y(k,2)) && (Y(k,2) < BinEdgesHigh2(i))
            P2(k) = Valores2(i);
        end
    end
    for i=1:size(Valores3,2)
        if (BinEdgesLow3(i) <= Y(k,3)) && (Y(k,3) < BinEdgesHigh3(i))
            P3(k) = Valores3(i);
        end
    end
end
lambda = 0.0000001;
P1 = lambda+P1;
P2 = lambda+P2;
P3 = lambda+P3;
N = P1+P2+P3;
ML = [P1,P2,P3]./N;
[confidence, ML_predict] = max(ML');
save 'ML.mat' ML
Test_predict_ML = ML_predict'-1;
save 'Test_predict_ML.mat' Test_predict_ML

**Comment by aldipiroli** [comment]
Thanks alot for the answer! 
Ah ok, yeah that makes sense.

I am trying to reimplement the method in python, so I want to be sure I got the correct general idea of it. 

1. First we perform inference on the training set using a detector `D` that predicts `C` total classes. Suppose that the detector predicts `n_total_objects`. What we will have an array of `total_logits = n_total_objects x C` raw output logits.
2. Using the ground truth labels, we select from `total_logits` all the true positives logits. Let's call it `logits_tp`.
3. Using  `logits_tp` we then compute, for each one of the `C` classes a normalized histogram distribution (`h_c`).

Now, during inference I can do the following:

4. I perform inference and get the raw output logits `pred_logits = n_predicted_objects x C`.
5. Now, for each prediction in `pred_logits`, find the value of the hisogram relative to logit class. 
For example:
if I have 3 classes a single prediction could be `current_pred_logit = 1x3 = [2.3, 7.5, 1.1]`.  
I go and look into `h_c1` what is the bin value corresponding to 2.3. The same for c2, c3. Suppose the values are [0.1, 0.5, 0.3].
Then, the new values for the current prediction will be `current_pred_logit_new =  [0.1, 0.5, 0.3]`.

6. After the last step, I will get the P value described in the code. I can then add lambda and divide by the sum of P along the class axis. This means that for each value of P, the sum among the class axis should be 1.  Similar to doing a softmax.

8. This results in the ML_TP vector described in the code, which is now the new prediction output of the detector. 
In order to get the class for a prediction, I can simply take the maximum value along the class axis and use this as my new prediction probability.


Sorry for the long message. Please let me know If I got anything wrong. Much appreciated.

**Comment by gledsonmelotti** [comment]
> Thanks alot for the answer! Ah ok, yeah that makes sense.
> 
> I am trying to reimplement the method in python, so I want to be sure I got the correct general idea of it.
> 
> 1. First we perform inference on the training set using a detector `D` that predicts `C` total classes. Suppose that the detector predicts `n_total_objects`. What we will have an array of `total_logits = n_total_objects x C` raw output logits.
> 2. Using the ground truth labels, we select from `total_logits` all the true positives logits. Let's call it `logits_tp`.
> 3. Using  `logits_tp` we then compute, for each one of the `C` classes a normalized histogram distribution (`h_c`).
> 
> Now, during inference I can do the following:
> 
> 4. I perform inference and get the raw output logits `pred_logits = n_predicted_objects x C`.
> 5. Now, for each prediction in `pred_logits`, find the value of the hisogram relative to logit class.
>    For example:
>    if I have 3 classes a single prediction could be `current_pred_logit = 1x3 = [2.3, 7.5, 1.1]`.
>    I go and look into `h_c1` what is the bin value corresponding to 2.3. The same for c2, c3. Suppose the values are [0.1, 0.5, 0.3].
>    Then, the new values for the current prediction will be `current_pred_logit_new =  [0.1, 0.5, 0.3]`.
> 6. After the last step, I will get the P value described in the code. I can then add lambda and divide by the sum of P along the class axis. This means that for each value of P, the sum among the class axis should be 1.  Similar to doing a softmax.
> 7. This results in the ML_TP vector described in the code, which is now the new prediction output of the detector.
>    In order to get the class for a prediction, I can simply take the maximum value along the class axis and use this as my new prediction probability.
> 
> Sorry for the long message. Please let me know If I got anything wrong. Much appreciated.

Hello.You are correct. But, determining the parameters can be a tiring step.  I adjusted by trial and error.

**Comment by aldipiroli** [comment]
Thanks for the feedback!

**Comment by aldipiroli** [comment]
Hi,
just another couple of clarifications:
1) How do you deal with logits values that don't fall in any of the histogram bins that you derive? Do you set the value to 0 or to the lowest/highest bin value? Something else?

2) When selecting the true positives detections from the training logits, what criterion do you use to distinguish between a true positive and false positve?
Thanks

**Comment by gledsonmelotti** [comment]
> Hi, just another couple of clarifications:
> 
> 1. How do you deal with logits values that don't fall in any of the histogram bins that you derive? Do you set the value to 0 or to the lowest/highest bin value? Something else?
> 2. When selecting the true positives detections from the training logits, what criterion do you use to distinguish between a true positive and false positve?
>    Thanks
Hi. It is zero because I defined lambda in my equations (Bayes-paper). I use bounding boxes and IOU in Detection.

**Comment by aldipiroli** [comment]
Thanks again!
